### PR TITLE
Recycle entity indices for constant-time creation

### LINKED ISSUE #44: Constant-time entity creation
Right now the `create` function has `O(n)` time complexity with `n` being the number of entities already existing in the world.

```ts
export function create(world: World, signature: number = 0) {
    for (let i = 0; i < MAX_ENTITIES; i++) {
        if (!world.Signature[i]) {
            world.Signature[i] = signature;
            return i;
        }
    }
    throw new Error("No more entities available.");
}
```

This might still be good enough for most use-cases we target, but I'd like to look into ways to optimize this. @ooflorent suggested an extra array storing the indices of destroyed entities. Let's call it `World.Graveyard`. When a new entity is requested, we first check the graveyard for an available index. If the graveyard is empty, we would use `World.Signature.length` as the next available index.

This is a fairly simple optimization in terms of complexity and codesize. The performance wins are well into one order of magnitude faster. However, I'm concerned about making entity creation dependent on the order of destroying entities. I think it might cause hard-to-debug bugs in `sys_transform` for example, which relies on the fact that parent entities are always updated before the children.

### PR DESCRIPTION
**Pull request by stasm**
Based on #51.
Fixes #44.

### CODE REVIEW

File: core/entity.ts [hunk 1]
```diff
@@ -3,16 +3,18 @@ import {transform} from "./components/com_transform.js";
...
+    }
+
+    if (world.Signature.length > 10000) {
+        throw new Error("No more entities available."); // DEBUG
     }
-    throw new Error("No more entities available.");
+
+    let entity = world.Signature.length;
+    world.Signature.push(0);
+    return entity;
```
Comment (ooflorent) [inline core/entity.ts:17]: This could be simplified:
```suggestion
    return world.Signature.push(0) - 1;
```

File: core/entity.ts [hunk 2]
```diff
@@ -22,6 +24,7 @@ export function destroy_entity(world: World, entity: Entity) {
         }
     }
     world.Signature[entity] = 0;
+    world.Graveyard.push(entity);
```
Comment (ooflorent) [inline core/entity.ts:27]: Maybe you could use a check (in debug mode) to make sure that an entity isn't pushed twice.

File: Animate/world.ts [hunk 3]
```diff
@@ -26,8 +26,9 @@ export const enum Has {
 }
 
 export class World {
-    // Component flags
     Signature: Array<number> = [];
+    Graveyard: Array<number> = [];
```
Comment (ooflorent) [inline Animate/world.ts:30]: I think it should be more specific:
```suggestion
    Graveyard: Array<Entity> = [];
```

File: core/entity.ts (hunk 1 above)
Comment (stasm) [reply core/entity.ts:17]: I'm not sure I'd call it _simpler_, but it's definitely shorter ;)

File: core/entity.ts (hunk 2 above)
Comment (stasm) [reply core/entity.ts:32]: Good idea, done!

**Pull request merged**